src/pke/lib/scheme/ckksrns/ckksrns-cryptoparameters.cpp

```diff
@@ -81,38 +81,54 @@ void CryptoParametersCKKSRNS::PrecomputeCRTTables(KeySwitchTechnique ksTech, Sca
         }
     }
 
-    // Pre-compute scaling factors for each level (used in EXACT scaling technique)
+    // Pre-compute scaling factors for each level (used in FLEXIBLE* scaling techniques)
     if (m_scalTechnique == FLEXIBLEAUTO || m_scalTechnique == FLEXIBLEAUTOEXT) {
         m_scalingFactorsReal.resize(sizeQ);
-        m_scalingFactorsReal[0] = moduliQ[sizeQ - 1].ConvertToDouble();
-
-        if (extraBits == 0) {
-            for (uint32_t k = 1; k < sizeQ; k++) {
-                double prevSF           = m_scalingFactorsReal[k - 1];
-                m_scalingFactorsReal[k] = prevSF * prevSF / moduliQ[sizeQ - k].ConvertToDouble();
-                double ratio            = m_scalingFactorsReal[k] / m_scalingFactorsReal[0];
-
-                if (ratio <= 0.5 || ratio >= 2.0)
-                    OPENFHE_THROW(
-                        "CryptoParametersCKKSRNS::PrecomputeCRTTables "
-                        "- FLEXIBLEAUTO cannot support this "
-                        "number of levels in this parameter setting. Please use "
-                        "FIXEDMANUAL.");
-            }
+
+        if ((sizeQ == 1) && (extraBits == 0)) {
+            // mult depth = 0 and FLEXIBLEAUTO
+            // when multiplicative depth = 0, we use the scaling mod size instead of modulus size
+            // Plaintext modulus is used in EncodingParamsImpl to store the exponent p of the scaling factor
+            m_scalingFactorsReal[0] = pow(2, GetPlaintextModulus());
+        }
+        else if ((sizeQ == 2) && (extraBits > 0)) {
+            // mult depth = 0 and FLEXIBLEAUTOEXT
+            // when multiplicative depth = 0, we use the scaling mod size instead of modulus size
+            // Plaintext modulus is used in EncodingParamsImpl to store the exponent p of the scaling factor
+            m_scalingFactorsReal[0] = moduliQ[sizeQ - 1].ConvertToDouble();
+            m_scalingFactorsReal[1] = pow(2, GetPlaintextModulus());
         }
         else {
-            m_scalingFactorsReal[1] = moduliQ[sizeQ - 2].ConvertToDouble();
-            for (uint32_t k = 2; k < sizeQ; k++) {
-                double prevSF           = m_scalingFactorsReal[k - 1];
-                m_scalingFactorsReal[k] = prevSF * prevSF / moduliQ[sizeQ - k].ConvertToDouble();
-                double ratio            = m_scalingFactorsReal[k] / m_scalingFactorsReal[1];
-
-                if (ratio <= 0.5 || ratio >= 2.0)
-                    OPENFHE_THROW(
-                        "CryptoParametersCKKSRNS::PrecomputeCRTTables "
-                        "- FLEXIBLEAUTO cannot support this "
-                        "number of levels in this parameter setting. Please use "
-                        "FIXEDMANUAL.");
+            m_scalingFactorsReal[0] = moduliQ[sizeQ - 1].ConvertToDouble();
+
+            if (extraBits == 0) {
```

Comment:
this if statement may be re-written to avoid redundancies

Reply:
My biggest concern here is the readability of the code. The underlying logic is very complicated (and I made it more complicated with the fix). I believe you are referring to the prior branching for mult dept > 0. If you can find a way to do compress it and still preserve the clarity, please feel free to do it. I will need to review the revised branches to make sure the logic is preserved.

Reply:
It seems that based on the value of extraBits, we can set k0 = 1 or k0 = 2 and m_scalingFactorsReal[1], then execute outside the if the for loop starting from k0.